### PR TITLE
ESMValCore Preprocessor: fixing unmatching attributes in pairwise concatenation.

### PR DESCRIPTION
**Pull request by francesco-cmcc**
Hi everybody.
I'm submitting this pull request to solve an issue encountered in regional dataset preprocessing.
When a dataset is composed by multiple files, the esmvalcore's preprocessor concatenates the netcdfs by a pairwise concatenation (`_by_two_contatenation` in `esmvalcore/preprocessor/_io.py`). 
For instance if our dataset is splitted in 3 files `file_1994.nc`, `file_1995.nc` and` file_1996.nc`, the preprocessor concatenates the dataset in 2 steps:

1. Concatenate  `file_1994.nc` and `file_1995.nc` into `file_1994_1995.nc`
2. Concatenate `file_1994_1995.nc` and `file_1996.nc` into `file_1994_1995_1996.nc`

Sometimes an attribute (or type) unmatching between `file_1994_1995.nc` and `file_1996.nc` could happen. This could be due to several reasons either iris backend or an error in a dataset. To fix this potential issue I added an `_equalize_attributes` fuction wich tries to equalize the attributes (and types) between each pair of cubes taking the first cube as reference.

I encountered this issue in EUROCORDEX preprocessing. If found a type unmatching in CCLM EUROCORDEX datasets. This issue is still being discussed in this pull request #1068, here the problem seems to be solved in a more accurate way. Anyway I decided  to share the code by this pull requests beacause my solution could be more generic.

Cheers
Francesco 

## Description

Closes #issue_number

Link to documentation:

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [ ] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [ ] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [ ] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [ ] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [ ] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [ ] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [ ] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [ ] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [ ] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

File: esmvalcore/preprocessor/_io.py [hunk 1]
```diff
@@ -461,3 +470,84 @@ def _concatenate_overlapping_cubes(cubes):
                 raise ex
 
     return cubes
+
+def _equalise_attributes(cubes_list):
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_io.py:474]: there is already an `iris` function available, funnily enough named the same, [`equalise_attributes`](https://scitools-iris.readthedocs.io/en/latest/generated/api/iris/util.html#iris.util.equalise_attributes)

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
many thanks for spotting this and suggesting a fix @francesco-cmcc :beer: Two things before a more ind epth review - please have a look at the builtin iris function `wqualise_attributes` and second, could you please open a direct Pull Request to the esmvalcore rpo, rather than a fork? It's easier for me to checkout the branch than the fork, cheers!

**Comment by francesco-cmcc** [comment]
Hi @valeriupredoi, I'm unable to create a branch directly in the ESMValgroup's main repository. 
Maybe, do I need some permission as contributor?

**Comment by valeriupredoi** [comment]
> Hi @valeriupredoi, I'm unable to create a branch directly in the ESMValgroup's main repository.
> Maybe, do I need some permission as contributor?
> I going to check `equalise_attributes` from iris documentation, thanks ;)

I have just added you to the broad Dev team, after you accept the invite you will be able to create PR's from branches branched straight off `main`, let me know if you still have issues :+1:

**Comment by valeriupredoi** [comment]
@francesco-cmcc could you please email me and @axel-lauer - Axel will send you the official contributors' email :beer:

**Comment by francesco-cmcc** [comment]
> @francesco-cmcc could you please email me and @axel-lauer - Axel will send you the official contributors' email 🍺

Ok. I'm sending you an email.

**Comment by francesco-cmcc** [comment]
Good morning,
I create a branch also from the main repository https://github.com/ESMValGroup/ESMValCore/tree/fix_concatenation_regional_domain

I also gave a look to the source code of iris `equalize_attributes`. It has a similar behaviour, but It only removes the unmatching attributes in a list of cubes, nothing is done on dtypes.   ​

To give this name also to my function, maybe has been an unhappy idea. 😃 

Francesco

**Comment by bouweandela** [comment]
The dtypes could also be fixed as part of the automatic fixes of the CMOR check, data should have typically have dtype `float32` and coordinates and bounds `float64`.

**Comment by valeriupredoi** [comment]
good stuff @francesco-cmcc :+1: I agree with Bouwe here though - it's best to fix input files at the very top of the stack via the cmor fix tool, and then see if there are any other issues with attributes and use the iris function to equalize (note that we are already using it in places). The reason why I mentioned the iris function and its use is that we'd like to be able, as much as possible, to make use of the already existing iris API - saves us duplication (to certain degree) work, generalizes a solution and makes it available elsewhere, and it's well understood and tested via iris development chain. Of course, a lot of the times we can't, but I reckon we can here, by adding the dtype fix in cmor fixes :+1:

**Comment by francesco-cmcc** [comment]
Hi @valeriupredoi  and @bouweandela,
I agree with you. If there is a way to fix this without adding a new function to preprocessor, it will be good. 
So, if I have well understood, are you proposing to use cmor.fix to solve this issue? 
Do this module already support this type of fix? or should it be implemented?
Thanks
Francesco

**Comment by bouweandela** [comment]
@jvegasbsc and @sloosvel are our experts on CMOR checks and fixes. Could you advise @francesco-cmcc on this?

**Comment by sloosvel** [comment]
There is an entry for the `type` in the cmor tables. I guess a check against it's value could be added during the cmor checks in `check.py` for the data.

So something like:

```
    def check_data(self, logger=None):
         ....
        self._check_data_type()
        self._check_coords_data() # and call the data type check in here again for the coords
       ...
```

However, the type is not something that is read at the moment. It should be added in the `VariableInfo` class in `table.py`. I don't know if @jvegasbsc has a better idea though.

**Comment by schlunma** [comment]
@francesco-cmcc do you have any intentions to further work on this for v2.5? I agree with @bouweandela and @valeriupredoi that it would be better to implement this as a [fix](https://github.com/ESMValGroup/ESMValCore/tree/main/esmvalcore/cmor/_fixes).

**Comment by schlunma** [comment]
Postponed to v2.6, I think this needs further discussion.

**Comment by francesco-cmcc** [comment]
> @francesco-cmcc do you have any intentions to further work on this for v2.5? I agree with @bouweandela and @valeriupredoi that it would be better to implement this as a [fix](https://github.com/ESMValGroup/ESMValCore/tree/main/esmvalcore/cmor/_fixes).

Sorry guys. I can't implement it in the next version. I don't follow this project anymore.

**Comment by schlunma** [comment]
All right, no worries! I will close this now, please re-open if necessary.

**Pull request closed without merging**